### ISSUE TITLE
The a element in HTML5

### ISSUE THREAD

**Issue by bytestream** [xemlock/htmlpurifier-html5#51]
> The a element may be wrapped around entire paragraphs, lists, tables, and so forth, even entire sections, so long as there is no interactive content within (e.g. buttons or other links). 
_https://html.spec.whatwg.org/dev/text-level-semantics.html#the-a-element_

**Example:**

```html
<a><table></table></a>
```

becomes

```html
<a></a><table></table>
```

**Comment by bytestream** [comment]
Presumably a fairly simple fix?

```php
$def->addElement('a', 'Inline', 'Flow', 'Common');
```
_https://github.com/xemlock/htmlpurifier-html5/blob/master/library/HTMLPurifier/HTMLModule/HTML5/Text.php#L68_

**Comment by xemlock** [comment]
Hi @bytestream,
This is not a bug of this library, it's how PHP DOMDocument (used by HTMLPurifier as [the default lexer](https://github.com/ezyang/htmlpurifier/blob/master/library/HTMLPurifier/Lexer/DOMLex.php)) parses the HTML. If you switch lexer implementation to `DirectLex`, then the parsing works as expected.

```php
$config = HTMLPurifier_HTML5Config::createDefault();
$config->set('Core.LexerImpl', 'DirectLex');

// now <A> won't be autoclosed when a <TABLE> is encountered
```

**Comment by bytestream** [comment]
Hey, thanks for the quick reply.

It should be default functionality though, given this lib advertises as HTML5 support and an anchor with block level elements is supported in HTML5? DOMDocument doesn't support HTML5 so while that's used, this lib can never fully be HTML5 friendly.

What are the implications of switching to DirectLex? htmlpurifier docs seem to suggest DirectLex is the default PHP 4 lexer which is a bit old school...

**Comment by bytestream** [comment]
I'm not happy with DirectLex as a replacement based on these comments:

> Our in-house implementation of a parser. A pure PHP parser, DirectLex has absolutely no dependencies, making it a reasonably good default for PHP4.

> It is blazingly fast (for large documents, it performs twenty times faster than HTMLPurifier_Lexer_DirectLex)

Perhaps time to look at replacing PH5P with https://github.com/Masterminds/html5-php

**Comment by xemlock** [comment]
Unfortunately `DirectLex` is the only working solution for now. And PH5P, the other lexer shipped with HTMLPurifier, also [can't handle HTML5 properly](https://github.com/ezyang/htmlpurifier/issues/226).

Switching lexer to html5-php sounds like a proper solution. It's definitely doable, but requires more development time (which unfortunately I don't have at the moment).

**Comment by bytestream** [comment]
Would you accept that here or should I send it to htmlpurifier? It's pretty simple to replace PH5P with masterminds/html5-php and it seems to resolve the issue I've mentioned here.

**Comment by xemlock** [comment]
I'd gladly accept it here, thanks 🙂.  I would prefer, however, to keep the default lexer intact (`DOMLex`), at least for the 0.1.x versions.

**Comment by bytestream** [comment]
Sent :) but pending feedback for two issues posted to masterminds/html5

**Comment by xemlock** [comment]
HTML5 parsing issues resolved by #52. Closing.
